lib/ansible/module_utils/azure_rm_common.py

```diff
@@ -994,10 +994,14 @@ def __init__(self, auth_source='auto', profile=None, subscription_id=None, clien
                 except Exception as e:
                     self.fail("cloud_environment {0} could not be resolved: {1}".format(raw_cloud_env, e.message), exception=traceback.format_exc())
 
-        if self.credentials.get('subscription_id', None) is None and self.credentials.get('credentials') is None:
-            self.fail("Credentials did not include a subscription_id value.")
-        self.log("setting subscription_id")
-        self.subscription_id = self.credentials['subscription_id']
+        # Use Subscription ID if specified by the client instead of the one in CLI
+        if subscription_id is not None:
```

Comment:
where this  subscripiton_id from? it's dangerous to set subscription_id only since azure credential is a set of subscription id, service principal credential. current way of setting azure credential via either env or file or  az cli is working well, didn't  see the point of setting subscription id separately.

Reply:
I have projects splitted between subscriptions, for instance some servers in **subscriotion A** and some other servers in **subscription B**.

I need to run ansible with one inventory composed by multiple **yaml** files, each one with settings per subscription.

I want to run this to configure the servers:

```
ansible -i inventory playbook,yaml
```

Where inventory is a folder with 2 files:
* inventory/subscription_a.yml
* inventory/subscription_b.yml

With the previous dynamic inventory script I created wrappers to export the some environment variables, so I could force parts of inventory to use a subscription. But I'm afraid this is now impossible to do with the inventory plugin.

----

Also Azure CLI and env vars are bit dangerous if you have multiple subscriptions  similar environments. I saw people running the playbook against the wrong servers because they had selected the wrong subscription, forcing this in the inventory file/config is much more safe from my experience.

Reply:
The **subscripiton_id** will be setup in the yaml file.

Reply:
@yungezz Could you help give a feedback for the comment when you're available? Thanks!

Reply:
@yungezz Please re-check when you're available! Thanks!

Reply:
We have exactly the same use case and are running into this exact same problem. We have multiple subscriptions (for split billing and authorizations) and having to set environment variables or perform 'az account set -s' commands is very prone to human error.

The fact that hosts and related resources belong to specific azure subscriptions is mostly a detail that our operators aren't concerned with during their work.

Reply:
> it's dangerous to set subscription_id only since azure credential is a set of subscription id, service principal credential.

There is a difference between the credentials a user needs to have in order to work on some object and the the identification of the resource being created. As far as I understand the implications of this change it allows one to explicitly specify which under which `subscription_id` that resource is created.

eg. An availability set has an id like: `/subscriptions/<subscription_id>/resourceGroups/<resource_group>/providers/Microsoft.Compute/availabilitySets/<name>`. In this case there is a `subscription_id` but no service principal.

I think it's fair to argue that one should also be able to explicitly specify the credentials that need to be for that specific resource creation. Though a sensible fallback to the first matching credentials for that specific `subscription_id` would work in most cases.

Reply:
@yungezz Any udpate for this?